File: Backtracking/Combinations/Combinations.go
package main

import (
	"fmt"
)

func combine(n int, k int) [][]int {
	var res [][]int
	var comb []int
	backtrack(1, &comb, &res, n, k)
	return res
}

func backtrack(start int, comb *[]int, res *[][]int, n int, k int) {
	if len(*comb) == k {
		copyList := make([]int, len(*comb))
		copy(copyList, *comb)
		*res = append(*res, copyList)
		return
	}
	for num := start; num <= n; num++ {
		*comb = append(*comb, num)
		backtrack(num+1, comb, res, n, k)
		*comb = (*comb)[:len(*comb)-1]
	}
}

func main() {
	var t int
	fmt.Scan(&t)
	for i := 0; i < t; i++ {
		var n, k int
		fmt.Scan(&n, &k)
		result := combine(n, k)
		for _, nextList := range result {
			for _, x := range nextList {
				fmt.Print(x, " ")
			}
			fmt.Println()
		}
	}
}

File: Backtracking/Combinations/Combinations.cpp
#include <iostream>
#include <vector>
using namespace std;

void backtrack(int start, vector<int>& comb, vector<vector<int>>& res, int n,
               int k) {
  if (comb.size() == k) {
    res.push_back(comb);
    return;
  }
  for (int num = start; num <= n; ++num) {
    comb.push_back(num);
    backtrack(num + 1, comb, res, n, k);
    comb.pop_back();
  }
}

vector<vector<int>> combine(int n, int k) {
  vector<vector<int>> res;
  vector<int> comb;
  backtrack(1, comb, res, n, k);
  return res;
}

int main() {
  int t;
  cin >> t;
  for (int i = 0; i < t; ++i) {
    int n, k;
    cin >> n >> k;
    vector<vector<int>> result = combine(n, k);
    for (const auto& nextList : result) {
      for (const auto& x : nextList) {
        cout << x << " ";
      }
      cout << endl;
    }
  }
  return 0;
}
